shell/browser/win/dark_mode.cc

```diff
@@ -0,0 +1,197 @@
+// Copyright (c) 2020 Microsoft Inc. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE-CHROMIUM file.
+
+#include "shell/browser/win/dark_mode.h"
+
+#pragma comment(lib, "Uxtheme.lib")
+
+// This namespace contains code from
+// https://github.com/ysc3839/win32-darkmode/
+// governed by the MIT license and (c) Richard Yu
+namespace {
+
+// 1903 18362
+enum PreferredAppMode { Default, AllowDark, ForceDark, ForceLight, Max };
+
+bool g_darkModeSupported = false;
+bool g_darkModeEnabled = false;
+DWORD g_buildNumber = 0;
+
+enum WINDOWCOMPOSITIONATTRIB {
+  WCA_USEDARKMODECOLORS = 26  // build 18875+
+};
+struct WINDOWCOMPOSITIONATTRIBDATA {
+  WINDOWCOMPOSITIONATTRIB Attrib;
+  PVOID pvData;
+  SIZE_T cbData;
+};
+
+using fnSetWindowCompositionAttribute =
+    BOOL(WINAPI*)(HWND hWnd, WINDOWCOMPOSITIONATTRIBDATA*);
+fnSetWindowCompositionAttribute _SetWindowCompositionAttribute = nullptr;
+
+bool IsHighContrast() {
+  HIGHCONTRASTW highContrast = {sizeof(highContrast)};
+  if (SystemParametersInfoW(SPI_GETHIGHCONTRAST, sizeof(highContrast),
+                            &highContrast, FALSE))
+    return highContrast.dwFlags & HCF_HIGHCONTRASTON;
+  return false;
+}
+
+using fnDwmSetWindowAttribute = HRESULT(WINAPI*)(HWND, DWORD, LPCVOID, DWORD);
+fnDwmSetWindowAttribute _DwmSetWindowAttribute = {};
+
+void RefreshTitleBarThemeColor(HWND hWnd, bool dark) {
+  LONG ldark = dark;
+  if (g_buildNumber >= 20161) {
+    // DWMA_USE_IMMERSIVE_DARK_MODE = 20
+    _DwmSetWindowAttribute(hWnd, 20, &ldark, sizeof dark);
+    return;
+  }
+  if (g_buildNumber >= 18363) {
+    auto data = WINDOWCOMPOSITIONATTRIBDATA{WCA_USEDARKMODECOLORS, &ldark,
+                                            sizeof ldark};
+    _SetWindowCompositionAttribute(hWnd, &data);
+    return;
+  }
+  _DwmSetWindowAttribute(hWnd, 0x13, &ldark, sizeof ldark);
+}
+
+// helper to load symbols from a module
+template <typename P>
+bool Symbol(HMODULE h, P* pointer, const char* name) {
+  if (P p = reinterpret_cast<P>(GetProcAddress(h, name))) {
+    *pointer = p;
+    return true;
+  }
+  return false;
+}
+
+// helper to load symbols from a module
+template <typename P>
+bool Symbol(HMODULE h, P* pointer, int number) {
+  return Symbol(h, pointer, MAKEINTRESOURCEA(number));
+}
+
+bool CheckBuildNumber(DWORD buildNumber) {
+  return (buildNumber == 17763 ||  // 1809
+          buildNumber == 18362 ||  // 1903
+          buildNumber == 18363 ||  // 1909
+          buildNumber == 19041);   // 2004
+}
+
+void InitDarkMode() {
+  HMODULE hDwmApi = LoadLibrary(L"DWMAPI");
+  if (hDwmApi != 0) {
+    Symbol(hDwmApi, &_DwmSetWindowAttribute, "DwmSetWindowAttribute");
```

Comment:
Can you use `base::win::PEImage` instead of defining our own symbol import utility.

Reply:
refactored to use `base/` in 46ec6e8a50207a8c62a8e329d4e3131168c2095d